File: ios/LaneShadow/Views/Molecules/AppHeader.swift
import LaneShadowTheme
import SwiftUI

// MARK: - App Header Component

/**
 * App header molecule component
 *
 * Provides a consistent app header with title, optional subtitle, and action areas.
 * Following React Native component from react-native/components/ui/molecules/app-header.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - `theme.colors.surface.default` (background)
 *   - `theme.colors.onSurface.default` (title/subtitle text)
 * - Layout:
 *   - Height: 48-56pt typical
 *   - Padding: horizontal 16pt, vertical 12pt
 *   - Leading icon area: 24pt tappable
 * - Typography:
 *   - Title: 16pt semibold
 *   - Subtitle: 14pt regular with 0.7 opacity
 * - Elevation:
 *   - Shadow radius 4pt, y offset 2pt (when elevated)
 *
 * ## Parameters
 * - title: Header title text
 * - subtitle: Optional subtitle text
 * - leadingIcon: Optional left icon (back, menu, etc.)
 * - trailingContent: Optional right action buttons
 * - elevated: Whether to show shadow (default: false)
 * - onLeadingClick: Optional tap handler for leading icon
 */
public struct LSAppHeader: View {
    @Environment(\.theme) private var theme

    private let title: String
    private let subtitle: String?
    private let leadingIcon: AnyView?
    private let trailingContent: AnyView?
    private let elevated: Bool
    private let onLeadingClick: (() -> Void)?

    public init(
        title: String,
        subtitle: String? = nil,
        leadingIcon: AnyView? = nil,
        trailingContent: AnyView? = nil,
        elevated: Bool = false,
        onLeadingClick: (() -> Void)? = nil
    ) {
        self.title = title
        self.subtitle = subtitle
        self.leadingIcon = leadingIcon
        self.trailingContent = trailingContent
        self.elevated = elevated
        self.onLeadingClick = onLeadingClick
    }

    // MARK: - Body

    public var body: some View {
        HStack(alignment: .center) {
            // Leading icon area
            if let leadingIcon {
                Button(action: {
                    onLeadingClick?()
                }) {
                    leadingIcon
                        .frame(width: 24, height: 24)
                }
                .buttonStyle(PlainButtonStyle())
            } else {
                Spacer()
                    .frame(width: 24, height: 24)
            }

            Spacer()

            // Title and subtitle
            VStack(alignment: .leading, spacing: 2) {
                Text(title)
                    .font(.system(size: 16, weight: .semibold))
                    .foregroundStyle(theme.colors.onSurface.default)

                if let subtitle {
                    Text(subtitle)
                        .font(.system(size: 14, weight: .regular))
                        .foregroundStyle(theme.colors.onSurface.default.opacity(0.7))
                }
            }

            Spacer()

            // Trailing content area
            if let trailingContent {
                trailingContent
            } else {
                Spacer()
                    .frame(width: 24, height: 24)
            }
        }
        .frame(maxWidth: .infinity)
        .frame(height: 48)
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
        .background(theme.colors.surface.default)
        .if(elevated) { view in
            view.shadow(
                color: Color.black.opacity(0.1),
                radius: 4,
                y: 2
            )
        }
    }
}

// MARK: - Conditional View Modifier

private extension View {
    @ViewBuilder
    func `if`(
        _ condition: Bool,
        transform: (Self) -> some View
    ) -> some View {
        if condition {
            transform(self)
        } else {
            self
        }
    }
}

// MARK: - Preview

#Preview("AppHeader - Basic") {
    LSAppHeader(title: "Rides")
        .laneShadowTheme()
}

#Preview("AppHeader - With Subtitle") {
    LSAppHeader(
        title: "Rides",
        subtitle: "Near San Francisco"
    )
    .laneShadowTheme()
}

#Preview("AppHeader - With Leading Icon") {
    LSAppHeader(
        title: "Rides",
        leadingIcon: AnyView(
            Image(systemName: "chevron.left")
                .foregroundStyle(Color.black)
        ),
        onLeadingClick: {
            print("Back tapped")
        }
    )
    .laneShadowTheme()
}

#Preview("AppHeader - With Trailing Content") {
    LSAppHeader(
        title: "Rides",
        trailingContent: AnyView(
            HStack(spacing: 12) {
                Image(systemName: "magnifyingglass")
                Image(systemName: "ellipsis")
            }
            .foregroundStyle(Color.black)
        )
    )
    .laneShadowTheme()
}

#Preview("AppHeader - Complete") {
    LSAppHeader(
        title: "Rides",
        subtitle: "Near San Francisco",
        leadingIcon: AnyView(
            Image(systemName: "chevron.left")
                .foregroundStyle(Color.black)
        ),
        trailingContent: AnyView(
            HStack(spacing: 12) {
                Image(systemName: "magnifyingglass")
                Image(systemName: "ellipsis")
            }
            .foregroundStyle(Color.black)
        ),
        elevated: true,
        onLeadingClick: {
            print("Back tapped")
        }
    )
    .laneShadowTheme()
}

File: android/app/src/main/java/com/laneshadow/ui/components/molecules/AppHeader.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.RectangleShape
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme
import com.laneshadow.ui.components.atoms.ThemedText

/**
 * AppHeader title variants
 *
 * Defines typography variants for the header title.
 */
enum class AppHeaderTitleVariant {
    Default,
    Large,
}

/**
 * AppHeader molecule component
 *
 * Application header bar with title and optional actions.
 * Following React Native wrapper patterns from react-native/components/ui/app-header.tsx
 *
 * @param title Header title text
 * @param subtitle Optional subtitle text
 * @param leadingIcon Left icon composable (back, menu, etc.)
 * @param trailingActions Right action buttons composable
 * @param elevated Whether to show shadow elevation
 * @param onLeadingClick Callback for leading icon click
 * @param modifier Modifier for the header container
 * @param titleVariant Typography variant for title (Default or Large)
 */
@Composable
fun AppHeader(
    title: String,
    subtitle: String? = null,
    leadingIcon: @Composable (() -> Unit)? = null,
    trailingActions: @Composable (RowScope.() -> Unit)? = null,
    elevated: Boolean = false,
    onLeadingClick: (() -> Unit)? = null,
    modifier: Modifier = Modifier,
    titleVariant: AppHeaderTitleVariant = AppHeaderTitleVariant.Default,
) {
    val theme = LocalLaneShadowTheme.current

    // Container configuration
    val headerHeight = 56.dp
    val horizontalPadding = theme.space.lg
    val verticalPadding = 12.dp

    // Elevation shadow (2dp when elevated)
    val elevation = if (elevated) 2.dp else 0.dp

    // Build modifier
    val headerModifier = modifier
        .shadow(
            elevation = elevation,
            shape = RectangleShape,
        )

    Surface(
        modifier = headerModifier,
        color = theme.colors.surface.default,
        tonalElevation = if (elevated) 2.dp else 0.dp,
    ) {
        Row(
            modifier = Modifier
                .padding(horizontal = horizontalPadding, vertical = verticalPadding),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            // Leading icon section
            if (leadingIcon != null) {
                val leadingModifier = Modifier.then(
                    if (onLeadingClick != null) {
                        Modifier.clickable(onClick = onLeadingClick)
                    } else {
                        Modifier
                    }
                )
                Box(
                    modifier = leadingModifier,
                    contentAlignment = Alignment.Center,
                ) {
                    leadingIcon()
                }
            } else {
                // Spacer for alignment when no leading icon
                Box(modifier = Modifier.size(24.dp))
            }

            // Title and subtitle section (center)
            Row(
                modifier = Modifier.weight(1f),
                horizontalArrangement = if (leadingIcon == null && trailingActions == null) {
                    Arrangement.Center
                } else {
                    Arrangement.Start
                },
                verticalAlignment = Alignment.CenterVertically,
            ) {
                if (subtitle != null) {
                    // Title + subtitle layout
                    Row(
                        horizontalArrangement = Arrangement.Start,
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        // Title
                        ThemedText(
                            text = title,
                            modifier = Modifier,
                        )

                        // Subtitle
                        ThemedText(
                            text = " • $subtitle",
                            color = theme.colors.onSurface.default.copy(alpha = 0.7f),
                            modifier = Modifier,
                        )
                    }
                } else {
                    // Title only
                    ThemedText(
                        text = title,
                        modifier = Modifier,
                    )
                }
            }

            // Trailing actions section
            if (trailingActions != null) {
                Row(
                    horizontalArrangement = Arrangement.End,
                    verticalAlignment = Alignment.CenterVertically,
                    content = trailingActions,
                )
            } else {
                // Spacer for alignment when no trailing actions
                Box(modifier = Modifier.size(24.dp))
            }
        }
    }
}
